Notify the cloud about planned disconnections

sergeuz opened the pull request:
### Problem

This PR implements a mechanism for graceful disconnection from the cloud on the following events:

- System reset (as part of the firmware update process or via `System.reset()`, `System.dfu()` and other APIs).
- Disconnecting via `Particle.disconnect()`.
- Shutting down network interfaces.
- Entering the listening mode.
- Entering one of the sleep modes.

Upon any of the above events, the device first sends a Goodbye message to the cloud, waits for the message to be received by the cloud, and then proceeds to close the connection with the cloud. On UDP platforms the message is sent as a confirmable message and needs to be acknowledged by the server. On TCP platforms the delivery of the message is ensured by half-closing the socket and waiting for the server to close the connection.

The Goodbye message carries the information about the disconnection reason, system reset reason and duration of the sleep (if applicable). The format of the data is described [here](https://docs.google.com/document/d/1bM8oy2l6TrWtLE0H7dQKWNTiEuyVyayevTR7NWwfQyk).

**Important:** This PR introduces the following changes in the current Device OS behavior:

- Device OS now completely disconnects from the cloud when going into sleep. Previously, it was leaving the comms layer and cloud socket in whatever state they were prior to entering the sleep mode.
- By default, `System.reset()` and other APIs no longer reset the device immediately if it's connected to the cloud.

This PR also refactors buffer appender classes (`BufferAppender`, `BufferAppender2`), so that there's only one `BufferAppender` class.

### Steps to Test

- Run unit tests.
- Run the `app/cloud_disconnect` test app.

### References

- [ch34913]

m-mcgowan commented on system/src/system_mode.cpp:
```diff
@@ -105,3 +136,38 @@ spark::feature::State system_thread_get_state(void*)
+int system_reset(unsigned mode, unsigned reason, unsigned value, unsigned flags, void* reserved) {
+    unsigned defaultReason = RESET_REASON_NONE;
+    switch (mode) {
+    case SYSTEM_RESET_MODE_NORMAL:
```
It's certainly cleaner, but I am wondering if the overhead of mapping values is worth it, compared to simply using the reset reason as the reset mode.

sergeuz replied:
The idea was to clearly separate reset modes (i.e. _how_ we want to reset) and reset reasons (_why_ we want to reset), so that it is possible to perform a specific type of a reset, e.g. a DFU reset, and still use whichever reason code would be useful for diagnostic purposes along with it.

m-mcgowan commented on system/src/system_network_manager_api.cpp:
```diff
@@ -401,7 +401,7 @@ void manage_network_connection() {
     if (SPARK_WLAN_RESET || SPARK_WLAN_SLEEP) {
         auto wasSleeping = SPARK_WLAN_SLEEP;
-        cloud_disconnect();
+        cloud_disconnect(CLOUD_DISCONNECT_GRACEFULLY, CLOUD_DISCONNECT_REASON_NETWORK_DISCONNECT);
```
I feel we should distinguish between user specified network disconnects, unanticipated network disconnects and the network watchdog disconnects. This will help us know whether the disconnect is intentional or not.

sergeuz replied:
:+1: I will need to add a new field to the Goodbye message and update its handler in Device Service accordingly.

sergeuz replied:
I added a new field that contains network disconnection reason to Goodbye message and updated the system cloud API accordingly. It should be noted that, currently, the only network disconnection reason that can be reported via Goodbye message is `NETWORK_DISCONNECT_REASON_USER`, since, normally, the system itself doesn't trigger a network disconnection, and when it does (e.g. to recover from recurring cloud connection errors), it's not connected to the cloud anyway.

m-mcgowan commented on test/unit_tests/services/varint.cpp:
```diff
@@ -0,0 +1,227 @@
+
+using namespace particle;
+
+TEST_CASE("encodeUnsignedVarint()") {
```
👏

m-mcgowan commented on user/tests/app/cloud_disconnect/cloud_disconnect.cpp:
```diff
@@ -0,0 +1,128 @@
+        break;
+    }
+    case Step::NETWORK_SLEEP: {
+        // FIXME: This mode seems to be broken on Gen 3 platforms
```
Please add a CH issue with more details if we don't yet have one.

sergeuz replied:
:+1:

m-mcgowan commented on user/tests/app/cloud_disconnect/cloud_disconnect.cpp:
```diff
@@ -0,0 +1,128 @@
+
+} // unnamed
+
+void setup() {
```
👏 👏

sergeuz commented on communication/inc/spark_protocol_functions.h:
```diff
@@ -225,16 +225,40 @@ int spark_protocol_post_description(ProtocolFacade* protocol, int desc_flags=par
+ */
+typedef struct spark_disconnect_command {
+    uint16_t size; ///< Size of this structure.
+    unsigned disconnect_reason; ///< Disconnection reason (a value defined by the `cloud_disconnect_reason` enum).
```
It's easy to change the size of an enum inadvertently, so I try to avoid using them in dynalib APIs

sergeuz commented on communication/src/lightssl_protocol.cpp:
```diff
@@ -22,20 +22,37 @@
-    result = this->wait_confirmable();
-    break;
-  case ProtocolCommands::TERMINATE:
+  case ProtocolCommands::DISCONNECT: {
```
Yeah, I saw that change in your PR. I don't quite understand the need for that mixin, given that we can simply move `wait_confirmable()` and other methods to `Protocol` and make them virtual. As for me, relations between all the classes in the comms library are already complex enough to introduce a new layer of indirection :man_shrugging:

sergeuz commented on system/src/main.cpp:
```diff
@@ -379,7 +379,7 @@ extern "C" void HAL_SysTick_Handler(void)
         {
             //Reset is the only way now to recover from stuck OTA update
-            HAL_Core_System_Reset_Ex(RESET_REASON_UPDATE_TIMEOUT, 0, nullptr);
+            HAL_Core_System_Reset_Ex(RESET_REASON_WATCHDOG, 0, nullptr);
```
Erm, I'm not sure :smile: I might have thought this reset doesn't deserve its own reason code :man_shrugging:

On second thought, how about we introduce a new code, e.g. `RESET_REASON_RECOVERY` and further categorize that reset reason using the [additional data](https://github.com/particle-iot/device-os/blob/2896ead2f717e5cdea2f733e502593064ee8c125/hal/src/nRF52840/core_hal.c#L67)? We use a similar approach with `RESET_REASON_PANIC`, and this is not the only place in the code where a system reset is performed as a last resort for recovery.

sergeuz replied:
I reverted this change

sergeuz commented:
> opt-out: if the developer doesn't want to be sending disconnection reasons to the cloud, perhaps this can be opted out, e.g. via a system flag or feature.

How about we instead add a feature flags that disables graceful disconnections altogether? I don't see much sense in disabling Goodbye messages alone, but still waiting for other unacknowledged messages when disconnecting from the cloud

sergeuz commented:
I noticed that graceful cloud disconnection was implemented in 3 different places, 09ecec3c363555c9d57856bab4c734b4c1e9bf76 and 32b06be653f6b7dce0c0e9a76d631ce00c4fad94 address that

sergeuz commented:
> opt-out: if the developer doesn't want to be sending disconnection reasons to the cloud, perhaps this can be opted out, e.g. via a system flag or feature

@m-mcgowan It is now possible to disable graceful cloud disconnection either globally (via the `SYSTEM_FLAG_DISCONNECT_IMMEDIATELY` flag), or by passing the `RESET_NO_WAIT` flag to `System.reset()` and other methods that reset the device.

sergeuz commented:
> I was thinking to save data. But we can add it later if needed.

@m-mcgowan hm, yeah, that would make sense. I'm not entirely happy about the `SYSTEM_FLAG_DISCONNECT_IMMEDIATELY` flag which I introduced in this PR. It feels that it acts too broadly and may actually cause problems on sleepy devices. Having something like `Particle.abort()` would be a much more flexible option.

How about I ditch that flag and introduce another one which would disable goodbye messages specifically?

m-mcgowan commented:
> How about I ditch that flag and introduce another one which would disable goodbye messages specifically?

I guess my main concern is that adding the goobye message adds two key behaviour changes:

1. disconnects take longer since they wait for the message to be ACKed
2. disconnects require more data

When introducing features, it's not always clear how these changes impact customers. 

On the basis of this, I'm suggesting adding two flags: one to disable the goodbye message entirely and another to disable waiting for it (so it's sent as a non-confirmable CoAP message.) The state of these flags will be communicated to the cloud so that it knows the device is intentionally not saying goodbye when leaving. Irish Goodbye :-)

sergeuz commented:
> I'm suggesting adding two flags: one to disable the goodbye message entirely and another to disable waiting for it (so it's sent as a non-confirmable CoAP message.)

It can be my dislike of overly configurable interfaces, but I'd keep it simple. If the overhead of this feature is not desirable, the users are free to disable it. Allowing it to be enabled but work unreliably feels like we're inventing extra work for ourselves.

> The state of these flags will be communicated to the cloud so that it knows the device is intentionally not saying goodbye when leaving. Irish Goodbye :-)

That's a good point. I'll add a relevant flag to the Hello message :+1:

sergeuz commented:
@m-mcgowan I added a system flag that enables/disables sending of Goodbye messages, as well as a Hello message flag that tells the cloud whether it should expect a Goodbye message from the device. Below is the summary of the [changes since your last review](https://github.com/particle-iot/device-os/pull/1899/files/b8eea372830cd18d3b43256c8ffaa3fa096eea85..0f7315986c38e60ba3304bc2a3730b16f984e13d):

- Protocol flags are now stored persistently in the session data. This allows the protocol implementation to detect changes in its configuration and trigger a new Hello message when the session is resumed.
- App state is no longer calculated as a CRC32 of other CRC32 checksums, instead, all fields comprising the app state are now compared directly. This reduces the chance of a collision.
- `SYSTEM_FLAG_DISCONNECT_IMMEDIATELY` was replaced with the `SYSTEM_FLAG_SEND_GOODBYE` flag (set to 1 by default).
- Fixed a bug with truncated describe messages introduced in this PR.

sergeuz commented on system/src/system_cloud_internal.cpp:
```diff
@@ -922,6 +922,10 @@ void Spark_Protocol_Init(void)
 
+        uint8_t disableGoodbye = 0;
+        system_get_flag(SYSTEM_FLAG_DISABLE_GOODBYE, &disableGoodbye, nullptr);
+        spark_protocol_set_connection_property(sp, particle::protocol::Connection::GOODBYE_ENABLED, !disableGoodbye, nullptr, nullptr);
```
Sounds like a good idea :+1:

sergeuz commented on communication/src/protocol.cpp:
```diff
@@ -300,13 +300,10 @@ int Protocol::begin()
 	{
 		// for now, unconditionally move the session on resumption
 		channel.command(MessageChannel::MOVE_SESSION, nullptr);
-		if (channel_flags & SKIP_SESSION_RESUME_HELLO) {
```
`SKIP_SESSION_RESUME_HELLO` is not used anywhere outside of this method, so there's no need to store that flag in the protocol instance. We also don't want to store it persistently in the session data.

sergeuz commented on communication/inc/protocol_defs.h:
```diff
@@ -111,10 +111,14 @@ enum DescriptionType {
 {
-    PING = 0,
-    FAST_OTA = 1
+    PING = 0, ///< Set keepalive interval.
```
It's a Doxygen syntax for putting docs after enum values: http://www.doxygen.nl/manual/docblocks.html#memberdoc

m-mcgowan replied:
TIL :-)

sergeuz commented on services/inc/appender.h:
```diff
@@ -88,11 +88,11 @@ class Appender {
     }
 
-    bool append(const char* str) {
+    bool append(const char* str) { // Deprecated, use appendString()
```
"Deprecated" was probably an overstatement :smile: It's just a more specific name like `appendInt16Le()`, `appendInt32Le()` and other new methods of this class.

avtolstoy commented on system/src/system_network_internal.h:
```diff
@@ -623,7 +623,9 @@ class ManagedNetworkInterface : public NetworkInterface
         if (SPARK_WLAN_STARTED)
         {
-            disconnect(NETWORK_DISCONNECT_REASON_NETWORK_OFF);
+            // It's expected that the system calls disconnect() before turning off the network to
```
See the comments about `disconnect()` calls in wiring.

sergeuz commented on communication/src/coap_channel.h:
```diff
@@ -568,13 +568,22 @@ class CoAPReliableChannel : public T
 		server.clear();
 		client.clear();
-		return channel::establish(flags, app_crc);
+		T::reset();
```
Fixed

sergeuz commented on communication/src/protocol.cpp:
```diff
@@ -354,9 +350,18 @@ int Protocol::begin()
+
 const auto HELLO_FLAG_OTA_UPGRADE_SUCCESSFUL = 1;
 const auto HELLO_FLAG_DIAGNOSTICS_SUPPORT = 2;
 const auto HELLO_FLAG_IMMEDIATE_UPDATES_SUPPORT = 4;
```
Fixed

sergeuz commented on services/inc/appender.h:
```diff
@@ -43,68 +44,69 @@ class Appender {
-    Appender* a = (Appender*)appender;
-    return a->append(data, length);
-}
+    bool appendInt8(uint8_t val) {
```
Fixed

sergeuz commented on system/inc/system_cloud.h:
```diff
@@ -276,16 +296,54 @@ void spark_cloud_flag_connect(void);
+    uint16_t size; ///< Size of this structure.
+    uint8_t flags; ///< Option flags (see `spark_cloud_disconnect_option_flag`).
+    uint8_t graceful; ///< Set to a non-zero value if graceful disconnection is enabled.
+    unsigned timeout; ///< Maximum time in milliseconds to wait for message acknowledgements.
```
Fixed

sergeuz commented on wiring/src/spark_wiring_network.cpp:
```diff
@@ -76,6 +76,7 @@ void NetworkClass::on() {
 }
 
 void NetworkClass::off() {
+    disconnect();
```
Fixed

sergeuz commented on wiring/inc/spark_wiring_wifi.h:
```diff
@@ -154,6 +154,7 @@ class WiFiClass : public NetworkClass
     }
 
     void off(void) {
+        disconnect();
```
Fixed

sergeuz commented on wiring/inc/spark_wiring_mesh.h:
```diff
@@ -145,6 +145,7 @@ class MeshClass : public NetworkClass, public MeshPublish {
     }
 
     void off() {
+        disconnect();
```
Fixed

sergeuz commented on wiring/inc/spark_wiring_cellular.h:
```diff
@@ -48,6 +48,7 @@ class CellularClass : public NetworkClass
         network_on(*this, 0, 0, NULL);
     }
     void off() {
+        disconnect();
```
Fixed

avtolstoy commented on wiring/inc/spark_wiring_cloud.h:
```diff
@@ -78,8 +78,8 @@ class CloudDisconnectOptions {
 
-    CloudDisconnectOptions& timeout(unsigned timeout);
-    unsigned timeout() const;
+    CloudDisconnectOptions& timeout(system_tick_t timeout);
```
Add chrono literals overload as well everywhere `system_tick_t` can be taken.

sergeuz replied:
:+1:
